sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataColumnsSuite.scala

```diff
@@ -0,0 +1,481 @@
+        }
+      }
+    }
+  }
+
+  Seq("true", "false").foreach { photonEnabled =>
+    Seq("true", "false").foreach { offHeapColumnVectorEnabled =>
+      withSQLConf("spark.sql.columnVector.offheap.enabled" -> offHeapColumnVectorEnabled,
+        "spark.databricks.photon.enabled" -> photonEnabled) {
+        metadataColumnsTest(s"read metadata with " +
+          s"offheap set to $offHeapColumnVectorEnabled, " +
+          s"photon set to $photonEnabled", schema) { (df, f0, f1) =>
+          // read all available metadata columns
+          checkAnswer(
+            df.select("name", "age", "id", "university",
+              METADATA_FILE_NAME, METADATA_FILE_PATH,
+              METADATA_FILE_SIZE, METADATA_FILE_MODIFICATION_TIME),
+            Seq(
+              Row("jack", 24, 12345L, "uom", f0(METADATA_FILE_NAME), f0(METADATA_FILE_PATH),
+                f0(METADATA_FILE_SIZE), f0(METADATA_FILE_MODIFICATION_TIME)),
+              Row("lily", 31, null, "ucb", f1(METADATA_FILE_NAME), f1(METADATA_FILE_PATH),
+                f1(METADATA_FILE_SIZE), f1(METADATA_FILE_MODIFICATION_TIME))
+            )
+          )
+
+          // read a part of metadata columns
+          checkAnswer(
+            df.select("name", "university", METADATA_FILE_NAME, METADATA_FILE_SIZE),
+            Seq(
+              Row("jack", "uom", f0(METADATA_FILE_NAME), f0(METADATA_FILE_SIZE)),
+              Row("lily", "ucb", f1(METADATA_FILE_NAME), f1(METADATA_FILE_SIZE))
+            )
+          )
+        }
+      }
+    }
+  }
+
+  //////////////////////////
+  // TEST METADATA STRUCT //
+  //////////////////////////
+
+  // has _metadata.file_name
+  val jsonData0 =
+    """
+      |{
+      | "name":"jack",
+      | "_metadata":{
+      |   "age":24,
+      |   "file_name":"jack.json"
+      | }
+      |}
+      |""".stripMargin
+  val jsonSchema0: StructType = new StructType()
+    .add(StructField("name", StringType))
+    .add(StructField("_metadata", new StructType()
+      .add(StructField("age", IntegerType))
+      .add(StructField("file_name", StringType))))
+
+  // no naming conflicts at all
+  val jsonData1 =
+    """
+      |{
+      | "name":"jack",
+      | "metadata":{
+      |   "age":24,
+      |   "file_name":"jack.json"
+      | }
+      |}
+      |""".stripMargin
+  val jsonSchema1: StructType = new StructType()
+    .add(StructField("name", StringType))
+    .add(StructField("metadata", new StructType()
+      .add(StructField("age", IntegerType))
+      .add(StructField("file_name", StringType))))
+
+  /**
+   * Create a JSON file named `fileName` with `data` under `dir` directory.
+   */
+  private def createJSONFile(data: String, dir: File, fileName: String): String = {
+    val dataFile = new File(dir, s"/$fileName")
+    dataFile.getParentFile.mkdirs()
+    val bytes = data.filter(_ >= ' ').getBytes
+    Files.write(dataFile.toPath, bytes)
+    dataFile.getCanonicalPath
+  }
+
+  test("test data schema has _metadata struct") {
```

Comment:
How does this test improve test coverage? Test with JSON?

Reply:
To test: when users have a struct called `_metadata`, the hidden file metadata struct and fields won't be resolved and cannot be selected - will throw `AnalysisException`.

The metadata columns won't overwrite users' columns

Reply:
It looks weird that we test both CSV and JSON, but not for the file format difference. Can we update the test cases here to always use JSON? I think it's better than CSV as it supports struct types.